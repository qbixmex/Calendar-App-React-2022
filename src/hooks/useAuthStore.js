import { useDispatch, useSelector } from "react-redux";
import calendarAPI from "../api/calendarAPI";
import { onChecking, onLogin, onLogout, onClearMessage } from "../store";

export const useAuthStore = () => {

  const { status, user, errorMessage } = useSelector( state => state.auth );

  const dispatch = useDispatch();

  const startLogin = async ({ email, password }) => {

    dispatch( onChecking() );

    try {

      const { data } = await calendarAPI.post('/auth', {
        email,
        password
      });

      localStorage.setItem('token', data.token );
      localStorage.setItem('token-init-date', new Date().getTime() );

      dispatch( onLogin({ name: data.name, uid: data.uid }) );

    } catch (error) {

      let errorMessage = error.response.data.msg;

      if (error.response.data.errors?.email) {
        errorMessage = error.response.data.errors.email.msg;
      }

      if (error.response.data.errors?.password) {
        errorMessage = error.response.data.errors.password.msg;
      }

      dispatch( onLogout(errorMessage) );

      setTimeout(() => {
        dispatch( onClearMessage() );
      }, 100);

    }
  };

  const startRegister = async ({ name, email, password }) => {
    dispatch( onChecking() );

    try {

      const { data } = await calendarAPI.post('/auth/new', {
        name,
        email,
        password
      });

      localStorage.setItem('token', data.token );
      localStorage.setItem('token-init-date', new Date().getTime() );

      dispatch( onLogin({ name: data.name, uid: data.uid }) );

    } catch (error) {

      let errorMessage = error.response.data?.msg || '';

      if (error.response.data.errors?.name) {
        errorMessage = error.response.data.errors.name.msg;
      }

      if (error.response.data.errors?.email) {
        errorMessage = error.response.data.errors.email.msg;
      }

      if (error.response.data.errors?.password) {
        errorMessage = error.response.data.errors.password.msg;
      }

      dispatch( onLogout(errorMessage) );

      setTimeout(() => {
        dispatch( onClearMessage() );
      }, 100);

    }
  };

  const checkAuthToken = async () => {

    const token = localStorage.getItem('token');

    if ( !token ) return dispatch( onLogout() );

    try {

      const { data } = await calendarAPI.get('/auth/renew');

      localStorage.setItem('token', data.token );
      localStorage.setItem('token-init-date', new Date().getTime() );

      dispatch( onLogin({ name: data.name, uid: data.uid }) );

    } catch (error) {
      localStorage.removeItem('token');
      localStorage.removeItem('token-init-date');
      return dispatch( onLogout() );
    }

  };

  const startLogout = () => {
    localStorage.removeItem('token');
    localStorage.removeItem('token-init-date');
    dispatch( onLogout() );
  }

  return {
    // Properties
    status,
    user,
    errorMessage,

    // Methods
    startLogin,
    startRegister,
    checkAuthToken,
    startLogout,
  };
};
