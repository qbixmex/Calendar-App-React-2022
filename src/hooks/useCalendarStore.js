import { useDispatch, useSelector } from "react-redux";
import { onSetActiveEvent, onAddNewEvent } from "../store";

export const useCalendarStore = () => {

  const dispatch = useDispatch();

  const {
    events,
    activeEvent,
  } = useSelector( state => state.calendar );

  const setActiveEvent = ( calendarEvent ) => {
    dispatch( onSetActiveEvent( calendarEvent ) );
  };

  const startSavingEvent = async ( calendarEvent ) => {
    // TODO: go to backend

    // TODO: ok
    if ( calendarEvent._id ) {
      // Updating existing event.
    } else {
      // Creating new event.
      dispatch( onAddNewEvent({ ...calendarEvent, _id: new Date().getTime() }) );
    }
  };

  return {
    // Properties
    events,
    activeEvent,

    // Methods
    setActiveEvent,
    startSavingEvent,
  };
};
