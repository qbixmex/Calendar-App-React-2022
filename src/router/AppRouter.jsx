import { Routes, Route, Navigate } from "react-router-dom";
import { LoginPage } from "../auth";
import { CalendarPage } from "../calendar";

export const AppRouter = () => {

  const mainRoutes = ( authStatus = 'not-authenticated' ) => {
    switch (authStatus) {
      case 'not-authenticated':
        return (<Route path="/auth/*" element={ <LoginPage /> } />);
      case 'authenticated':
        return (<Route path="*" element={ <CalendarPage /> } />);
      default:
        return;
    }
  };

  return (
    <Routes>
      { mainRoutes('authenticated') }
      <Route path="*" element={ <Navigate to="/auth/login" /> } />
    </Routes>
  );
};
