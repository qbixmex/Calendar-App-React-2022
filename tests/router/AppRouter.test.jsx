import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { useAuthStore } from '../../src/hooks/useAuthStore';
import { AppRouter } from '../../src/router/AppRouter';

jest.mock('../../src/hooks/useAuthStore');

jest.mock('../../src/calendar/pages/CalendarPage', () => ({
  CalendarPage: () => <h1>Calendar Page</h1>,
}));

describe('Test on AppRouter', () => {
  const mockCheckAuthToken = jest.fn()

  beforeEach( () => jest.clearAllMocks() );

  test('Should show with default values', () => {

    useAuthStore.mockReturnValue({
      status: 'checking',
      checkAuthToken: mockCheckAuthToken
    });

    const { container } = render( <AppRouter /> );

    const spinner = container.querySelector('#spinner');

    expect( spinner.innerHTML ).toContain('Loading');
    expect( mockCheckAuthToken ).toHaveBeenCalled();

  });

  test('Should render to login page if user is not authenticated', () => {

    useAuthStore.mockReturnValue({
      status: 'not-authenticated',
      checkAuthToken: mockCheckAuthToken
    });

    const { container, getByRole } = render(
      <MemoryRouter initialEntries={['/auth/login']}>
        <AppRouter />
      </MemoryRouter>
    );

    expect(getByRole('heading', { name: 'Login' })).toBeTruthy();
    expect(getByRole('heading', { name: 'Register' })).toBeTruthy();
    expect( container ).toMatchSnapshot();

  });

  test('Should render calendar page if user is authenticated', () => {

    useAuthStore.mockReturnValue({
      status: 'authenticated',
      checkAuthToken: mockCheckAuthToken
    });

    const { getByRole } = render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    );

    const h1 = getByRole('heading', { level: 1 })

    expect( mockCheckAuthToken ).toHaveBeenCalled();
    expect(h1.innerHTML).toBe('Calendar Page');

  });
});
