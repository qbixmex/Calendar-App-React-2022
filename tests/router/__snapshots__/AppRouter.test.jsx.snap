// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on AppRouter Should render to login page if user is not authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row g-4"
    >
      <div
        class="col-lg-6 login-form-1"
      >
        <h2>
          Login
        </h2>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              id="login-email"
              name="loginEmail"
              placeholder="Correo"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-4"
          >
            <input
              class="form-control"
              id="login-password"
              name="loginPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid"
          >
            <button
              class="btnSubmit"
              type="submit"
            >
              Login
            </button>
          </div>
        </form>
      </div>
      <div
        class="col-lg-6 login-form-2"
      >
        <h2>
          Register
        </h2>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              id="register-name"
              name="registerName"
              placeholder="Nombre"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              id="register-email"
              name="registerEmail"
              placeholder="Correo"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              id="register-password"
              name="registerPassword"
              placeholder="Contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-4"
          >
            <input
              class="form-control"
              id="register-password-2"
              name="registerPassword2"
              placeholder="Repita la contraseña"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid"
          >
            <button
              class="btnSubmit"
              type="submit"
            >
              Create Account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
